pkg/apis/pipeline/v1beta1/param_types.go

```diff
@@ -284,12 +290,48 @@ func validateParameterInOneOfMatrixOrParams(matrix []Param, params []Param) (err
 	return errs
 }
 
-func validateStringVariable(value, prefix string, stringVars sets.String, arrayVars sets.String) *apis.FieldError {
+// validateStringVariableP validates param variable usage in string value
+// This will allow referencing the whole object/array value. example:
+// params:
+// 	- name: arg
+// 		value: $(params.myObject[*])
+// see https://github.com/tektoncd/pipeline/issues/4879#issuecomment-1133120522 for background
+func validateStringVariableP(param Param, prefix string, paramNames sets.String, arrayVars sets.String, objectParamNameKeys map[string][]string) (errs *apis.FieldError) {
+	stringValue := param.Value.StringVal
+	singleVariableRegex := regexp.MustCompile(fmt.Sprintf("^\\$\\(%s\\.[_a-zA-Z][_a-zA-Z0-9.-]*\\[\\*\\]\\)$", prefix))
+
+	// if the provided param value is just $(params.myObject[*]) or $(params.myArray[*]), example:
+	// 	params:
+	// 	- name: arg
+	// 		value: $(params.myObject[*])
+	if singleVariableRegex.MatchString(stringValue) {
+		return errs.Also(substitution.ValidateVariableP(stringValue, prefix, paramNames).ViaFieldKey("params", param.Name))
+	}
+
+	// if the provided param value is string literal and/or contains multiple variables
+	// valid example: "$(params.myString) and another $(params.myObject.key1)"
+	// invalid example: "$(params.myString) and another $(params.myObject[*])"
```

Comment:
I'm a bit confused what's going on here. How is `validateVariableP` different from `validateStringVariable` and `validateStringVariableP`? Can any of these functions be renamed?

also, is the problem with the second example that it is a string which contains both a string param and and object param (without referencing the key)?

Reply:
> I'm a bit confused what's going on here. How is `validateVariableP` different from `validateStringVariable` and `validateStringVariableP`? Can any of these functions be renamed?

Thanks. The names were indeed a bit confusing. I changed the name and add comments. Please take a look. Basically, `validateStringVariableP` (now called `validateParamStringValue`) allows using the references to the whole array or object i.e. `$(params.myarray[*])`, or `$(params['myarray'][*])` whereas the existing `validateStringVariable` doesn't allow to have a reference that contains an array param name whatsoever.

> also, is the problem with the second example that it is a string which contains both a string param and and object param (without referencing the key)?

Exactly!